File: backend/app/services/streak_service.py
"""Learning-streak bookkeeping.

A "learning day" is recorded whenever the student completes a lesson (or
makes ≥50% progress on one). The streak is the count of consecutive
learning days ending today (or yesterday, while today is still pending).

Streak restores: if the student skips a day, they can spend one of their
monthly restores (default 4) to back-fill that day and keep the streak
alive. Restores are counted against the calendar month of `restored_at`.
"""

from __future__ import annotations

import uuid
from datetime import date, datetime, timedelta

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.config import settings  # MAX_STREAK_RESTORES_PER_MONTH
from app.models import LearningDay


def _utc_date() -> date:
    return datetime.utcnow().date()


def _month_start(d: date) -> date:
    """First day of the month containing `d` (for the restore counter)."""
    return d.replace(day=1)


async def record_learning_day(
    db: AsyncSession,
    user_id: uuid.UUID,
    day: date | None = None,
) -> LearningDay:
    """Idempotently record a learning day for `user_id` (default today)."""
    day = day or _utc_date()
    result = await db.execute(
        select(LearningDay).where(LearningDay.user_id == user_id, LearningDay.day == day)
    )
    existing = result.scalar_one_or_none()
    if existing is not None:
        # Already learned that day — no duplicate row.
        return existing
    row = LearningDay(user_id=user_id, day=day, is_restored=False)
    db.add(row)
    await db.commit()
    await db.refresh(row)
    return row


async def _load_rows(db: AsyncSession, user_id: uuid.UUID) -> list[LearningDay]:
    """Fetch every learning-day row for the user (used for all streak math)."""
    result = await db.execute(select(LearningDay).where(LearningDay.user_id == user_id))
    return list(result.scalars().all())


def _current_streak(learned: set[date], today: date) -> int:
    """Count consecutive learning days ending today (or yesterday while
    today is still pending — a streak isn't broken until the day passes)."""
    if today in learned:
        anchor = today
    elif (today - timedelta(days=1)) in learned:
        anchor = today - timedelta(days=1)
    else:
        # Neither today nor yesterday learned ⇒ streak has lapsed.
        return 0
    streak = 0
    d = anchor
    while d in learned:
        streak += 1
        d -= timedelta(days=1)
    return streak


def _longest_streak(learned: set[date]) -> int:
    """Longest run of consecutive days in the user's history."""
    if not learned:
        return 0
    best = 0
    current = 0
    prev: date | None = None
    for d in sorted(learned):
        # Reset the run when a gap appears, otherwise extend it.
        current = current + 1 if prev is not None and (d - prev).days == 1 else 1
        best = max(best, current)
        prev = d
    return best


def _restorable_day(learned: set[date], today: date) -> date | None:
    """Most recent skipped day (< today) that sits right after a learned day.

    Scanning at most 8 days back keeps restores limited to recent skips.
    """
    for offset in range(1, 9):
        d = today - timedelta(days=offset)
        if d in learned:
            continue
        # Only a day whose predecessor was learned can be back-filled
        # meaningfully (it slots into the streak chain).
        if (d - timedelta(days=1)) in learned:
            return d
    return None


def _restores_used_this_month(rows: list[LearningDay]) -> int:
    """Count of restored days in the current calendar month."""
    month = _month_start(datetime.utcnow().date())
    return sum(
        1
        for row in rows
        if row.is_restored and row.restored_at is not None and row.restored_at.date() >= month
    )


async def get_streak(db: AsyncSession, user_id: uuid.UUID) -> dict:
    """Assemble the full streak snapshot for the dashboard response."""
    rows = await _load_rows(db, user_id)
    learned = {row.day for row in rows}
    today = _utc_date()
    restorable = _restorable_day(learned, today)
    restores_used = _restores_used_this_month(rows)
    max_restores = settings.MAX_STREAK_RESTORES_PER_MONTH
    last = max(learned) if learned else None
    month = _month_start(today)
    return {
        "current_streak": _current_streak(learned, today),
        "longest_streak": _longest_streak(learned),
        "last_learning_day": last.isoformat() if last else None,
        "learned_today": today in learned,
        "days_this_month": sum(1 for row in rows if row.day >= month),
        "restores_used": restores_used,
        "restores_available": max(0, max_restores - restores_used),
        "max_restores_per_month": max_restores,
        "restorable_day": restorable.isoformat() if restorable else None,
        # Only eligible when there is a skippable day AND a restore left.
        "restore_eligible": restorable is not None and (max_restores - restores_used) > 0,
    }


async def restore_skipped_day(
    db: AsyncSession, user_id: uuid.UUID
) -> tuple[dict, str | None]:
    """Back-fill the most recent skipped day. Returns (streak_dict, error)."""
    rows = await _load_rows(db, user_id)
    learned = {row.day for row in rows}
    today = _utc_date()
    restorable = _restorable_day(learned, today)
    max_restores = settings.MAX_STREAK_RESTORES_PER_MONTH
    restores_used = _restores_used_this_month(rows)

    # Guard conditions: without them there is nothing to restore or no
    # budget left, so return the current state plus a human message.
    if restorable is None:
        return await get_streak(db, user_id), "No skipped day to restore."
    if restores_used >= max_restores:
        return await get_streak(db, user_id), "All restores for this month are used up."

    # Splice the skipped day back in, flagged as restored, and re-summarize.
    db.add(
        LearningDay(
            user_id=user_id,
            day=restorable,
            is_restored=True,
            restored_at=datetime.utcnow(),
        )
    )
    await db.commit()
    return await get_streak(db, user_id), None

File: backend-node/src/services/streakService.ts
/**
 * Learning-streak bookkeeping — mirrors backend/app/services/streak_service.py.
 *
 * A "learning day" is recorded when the student completes a lesson (or makes
 * ≥50% progress on one). The streak counts consecutive learning days ending
 * today (or yesterday while today is still pending). Restores let the student
 * back-fill a skipped day, limited per calendar month.
 */
import { config } from "../config.js";
import { db } from "../db.js";
import type { Row } from "../db.js";

export interface LearningDayRow extends Row {
  id: string;
  user_id: string;
  day: string; // YYYY-MM-DD
  is_restored: boolean;
  restored_at: string | null;
  created_at: string;
}

function utcDateStr(): string {
  return new Date().toISOString().slice(0, 10);
}

function monthStartStr(d: Date): string {
  return `${d.getUTCFullYear()}-${String(d.getUTCMonth() + 1).padStart(2, "0")}-01`;
}

export async function recordLearningDay(userId: string, day?: string): Promise<void> {
  const target = day ?? utcDateStr();
  const existing = await db.get<LearningDayRow>(
    `SELECT * FROM learning_days WHERE user_id = $1 AND day = $2`,
    [userId, target]
  );
  if (existing) return;
  await db.query(
    `INSERT INTO learning_days (id, user_id, day, is_restored, restored_at, created_at)
     VALUES ($1,$2,$3,FALSE,NULL,$4)`,
    [crypto.randomUUID(), userId, target, new Date().toISOString()]
  );
}

async function loadRows(userId: string): Promise<LearningDayRow[]> {
  return db.query<LearningDayRow>(`SELECT * FROM learning_days WHERE user_id = $1`, [userId]);
}

function currentStreak(learned: Set<string>, today: string): number {
  const todayMs = Date.parse(`${today}T00:00:00Z`);
  const fmt = (ms: number) => new Date(ms).toISOString().slice(0, 10);
  let anchor = today;
  if (!learned.has(today)) {
    const yest = fmt(todayMs - 86_400_000);
    if (learned.has(yest)) anchor = yest;
    else return 0;
  }
  let streak = 0;
  let d = Date.parse(`${anchor}T00:00:00Z`);
  while (learned.has(fmt(d))) {
    streak += 1;
    d -= 86_400_000;
  }
  return streak;
}

function longestStreak(learned: Set<string>): number {
  if (learned.size === 0) return 0;
  const days = [...learned].sort();
  let best = 0;
  let current = 0;
  let prevMs = 0;
  for (const d of days) {
    const ms = Date.parse(`${d}T00:00:00Z`);
    current = prevMs !== 0 && ms - prevMs === 86_400_000 ? current + 1 : 1;
    best = Math.max(best, current);
    prevMs = ms;
  }
  return best;
}

function restorableDay(learned: Set<string>, today: string): string | null {
  const todayMs = Date.parse(`${today}T00:00:00Z`);
  const fmt = (ms: number) => new Date(ms).toISOString().slice(0, 10);
  for (let offset = 1; offset <= 8; offset++) {
    const d = todayMs - offset * 86_400_000;
    const day = fmt(d);
    if (learned.has(day)) continue;
    const prev = fmt(d - 86_400_000);
    if (learned.has(prev)) return day;
  }
  return null;
}

function restoresUsedThisMonth(rows: LearningDayRow[]): number {
  const month = monthStartStr(new Date());
  return rows.filter(
    (row) => row.is_restored && row.restored_at && row.restored_at.slice(0, 10) >= month
  ).length;
}

export async function getStreak(userId: string): Promise<Record<string, unknown>> {
  const rows = await loadRows(userId);
  const learned = new Set(rows.map((r) => r.day));
  const today = utcDateStr();
  const restorable = restorableDay(learned, today);
  const restoresUsed = restoresUsedThisMonth(rows);
  const maxRestores = config.MAX_STREAK_RESTORES_PER_MONTH;
  const last = learned.size ? [...learned].sort().pop()! : null;
  const month = monthStartStr(new Date());

  return {
    current_streak: currentStreak(learned, today),
    longest_streak: longestStreak(learned),
    last_learning_day: last,
    learned_today: learned.has(today),
    days_this_month: rows.filter((r) => r.day >= month).length,
    restores_used: restoresUsed,
    restores_available: Math.max(0, maxRestores - restoresUsed),
    max_restores_per_month: maxRestores,
    restorable_day: restorable,
    restore_eligible: restorable !== null && maxRestores - restoresUsed > 0,
  };
}

/** Back-fill the most recent skipped day. Returns (streak, error|null). */
export async function restoreSkippedDay(
  userId: string
): Promise<[Record<string, unknown>, string | null]> {
  const rows = await loadRows(userId);
  const learned = new Set(rows.map((r) => r.day));
  const today = utcDateStr();
  const restorable = restorableDay(learned, today);
  const maxRestores = config.MAX_STREAK_RESTORES_PER_MONTH;
  const restoresUsed = restoresUsedThisMonth(rows);

  if (!restorable) return [await getStreak(userId), "No skipped day to restore."];
  if (restoresUsed >= maxRestores) {
    return [await getStreak(userId), "All restores for this month are used up."];
  }

  await db.query(
    `INSERT INTO learning_days (id, user_id, day, is_restored, restored_at, created_at)
     VALUES ($1,$2,$3,TRUE,$4,$4)`,
    [crypto.randomUUID(), userId, restorable, new Date().toISOString()]
  );
  return [await getStreak(userId), null];
}